src/Http/Routing/src/RouteEndpointBuilder.cs

```diff
@@ -12,15 +12,17 @@ namespace Microsoft.AspNetCore.Routing;
 /// </summary>
 public sealed class RouteEndpointBuilder : EndpointBuilder
 {
-    // TODO: Make this public as a gettable IReadOnlyList<Func<RouteHandlerContext, EndpointFilterDelegate, EndpointFilterDelegate>>.
-    // AddEndpointFilter will still be the only way to mutate this list.
-    internal List<Func<EndpointFilterFactoryContext, EndpointFilterDelegate, EndpointFilterDelegate>>? EndpointFilterFactories { get; set; }
-
     /// <summary>
     /// Gets or sets the <see cref="RoutePattern"/> associated with this endpoint.
     /// </summary>
     public RoutePattern RoutePattern { get; set; }
 
+    /// <summary>
+    /// Gets the collection of route handler filter factories associated with this endpoint.
+    /// </summary>
+    public IList<Func<EndpointFilterFactoryContext, EndpointFilterDelegate, EndpointFilterDelegate>> RouteHandlerFilterFactories { get; }
```

Comment:
See https://github.com/dotnet/aspnetcore/pull/42557, this is being moved to the base and the name should be changed. This name is redundant.

Reply:
I did see that. I figured that was just you prototyping in that PR though. Where was the discussion for this? Is there an API proposal? This wasn't mentioned in #42592 in the weeks following its approval.

I agree that removing `RouteHandler` from the name and just calling it `FilterFactories` is better. I do not think this should be moved to base type unless we also add the `RoutePattern` to the base type as well which I don't see happening because of the shadowing. See the following:

https://github.com/dotnet/aspnetcore/blob/8fe74e4d3ef2d26f58e3a2dfef90c64233941cfe/src/Http/Http.Extensions/src/RequestDelegateFactoryOptions.cs#L55-L57

I know you think this creates too tight of a coupling between routing and RDF, but itis very limiting for filters to not have access to the route pattern. `EndpointBuilders` without a `RoutePattern` [don't really exist](https://grep.app/search?q=%3A%20EndpointBuilder&filter[lang][0]=C%23). They're all `RouteEndpointBuilders` because who has time to reimplement route matching without `RoutePatterns`?

Maybe we should unseal `RouteEndpointBuilder`. It seems that this would help [the one non-test, non-inert] `EndpointBuilder`, [ServiceComposer's CompositionEndpointBuilder](https://github.com/ServiceComposer/ServiceComposer.AspNetCore/blob/9eba39a0ab7c7a94751ba6b3e065d5558bf57510/src/ServiceComposer.AspNetCore/CompositionEndpointBuilder.cs) which really want to be a `RouteEndpointBuilder` considering it has a `RoutePattern` it cannot expose and it returns a `RouteEndpoint` from `Build()`. @mauroservienti

Reply:
When I made that comment in RequestDelegateFactoryOptions and when leaving my last comment minutes ago, I forgot that the direction of the Routing -> Http.Extension dependency makes it impossible to expose `RoutePattern`s to filters without some crazy type forwarding. That is sad. Maybe we'll just expose it via metadata 😆

I still think we should just unseal `RouteEndpointBuilder` and leave `FilterFactories` there.

Reply:
I don't see why we would leave it on the derived type when the type got renamed to Endpoint* and not Route*. It has nothing to do with routing at this point.

Reply:
Now that I realize that the RoutePattern cannot be exposed to the filters because of the layering, I feel less strongly about it being in RouteEndopintBuilder. I'll update the API proposals to match what's in https://github.com/dotnet/aspnetcore/pull/42557 before our normal Monday API review.

I wish the routing stuff was lower level though. I still haven't seen a use case for EndpointBuilders or Endpoints that are not a RouteEndopintBuilders or RouteEndpoints. Having to downcast all the time is annoying. Are there any examples of this? Or is it more of a just-in-case thing? I realize you could in theory do your own matching based on something other than the RoutePattern. I just haven't seen it.

Reply:
The general use of `EndpointBuilder` as a context object makes sense to me. Is this use of builders-as-context-objects common anywhere else? The name can make it hard to understand that the builder is used by RDF and filters for accessing route properties, and not for actually modifying conventions on the route endpoint builder.

Reply:
> Is this use of builders-as-context-objects common anywhere else?

I'd consider it more of a builders-in-context-objects, but I don't know of another example in ASP.NET. Contexts and builders as separate parameters are more common, but that's for callbacks that are specifically intended to add to the builder.

It might actually make sense to change `IEndpointMetadataProvider.PopulateMetadata(EndpointMetadataContext)` to `IEndpointMetadataProvider.PopulateMetadata(MethodInfo, EndpointBuilder)` and `IEndpointParameterMetadataProvider.PopulateMetadata(EndpointParameterMetadataContext)` to `IEndpointParameterMetadataProvider.PopulateMetadata(ParameterInfo, EndpointBuilder)` and get rid of the new contexts entirely.

For filters, we could change the factory type from `Func<EndpointFilterFactoryContext, EndpointFilterDelegate, EndpointFilterDelegate>` to `Func<MethodInfo, EndopintBuilder, EndpointFilterDelegate, EndpointFilterDelegate>` and get rid of the `EndpointFilterFactoryContext`.

In the filter case, I like the flexibility of the context though, and the primary purpose of filters is probably not going to be to update metadata on the `EndpointBuilder`. This is the whole point of the meta data providers which is why I support removing the contexts there. @DamianEdwards 

> The name can make it hard to understand that the builder is used by RDF and filters for accessing route properties, and not for actually modifying conventions on the route endpoint builder.

I don't think this will be confusing because it's not like we're exposing an `IEndpointConventionBuilder` which has a bunch of extension methods targeting it that add conventions. The `EndpointBuilder` does make it easy to modify metadata which many conventions do, but filters and metadata providers are still going to have to add the metadata manually rather than use extension methods just like today.